### ISSUE TITLE
Additional features in `structured_negative_sampling`

### ISSUE THREAD

**Issue by paulagd** [rusty1s/pytorch_geometric#2197]
From now on, we recommend using our discussion forum (https://github.com/rusty1s/pytorch_geometric/discussions) for proposing new features.

## 🚀 Feature

def custom_negative_sampling(edge_index, num_nodes, num_ng_samples=1):
    r"""Samples a negative edge :obj:`(i,k)` for every positive edge
    :obj:`(i,j)` in the graph given by :attr:`edge_index`, and returns it as a
    tuple of the form :obj:`(i,j,k)`.

    Args:
        edge_index (LongTensor): The edge indices.
        num_nodes (int, optional): The number of nodes, *i.e.*
            :obj:`max_val + 1` of :attr:`edge_index`. (default: :obj:`None`)

    :rtype: (LongTensor, LongTensor, LongTensor)
    """
    i, j = edge_index.to('cpu')
    idx_1 = i * num_nodes + j
    k_list = []
    for t in range(num_ng_samples):
        k = torch.randint(j.min(), num_nodes, (i.size(0),), dtype=torch.long)
        idx_2 = i * num_nodes + k

        mask = torch.from_numpy(np.isin(idx_2, idx_1)).to(torch.bool)
        rest = mask.nonzero(as_tuple=False).view(-1)
        while rest.numel() > 0:  # pragma: no cover
            tmp = torch.randint(num_nodes, (rest.numel(),), dtype=torch.long)
            idx_2 = i[rest] * num_nodes + tmp
            mask = torch.from_numpy(np.isin(idx_2, idx_1)).to(torch.bool)
            k[rest] = tmp
            rest = rest[mask.nonzero(as_tuple=False).view(-1)]
        k_list.append(k)
    k_list = [aux.to(edge_index.device) for aux in k_list]

    return edge_index[0], edge_index[1], torch.stack(k_list)

## Motivation

Regarding the sampling in bi-parted graphs, I have found a useful function in your docs: [structured_negative_sampling](https://pytorch-geometric.readthedocs.io/en/latest/modules/utils.html#torch_geometric.utils.structured_negative_sampling) that aims to sample negative pair nodes. I have implemented a function in order to sample just second type of nodes (as in for example items for a given user) as k. Also I have added the option in order to sample `num_ng_samples` times. Just sending it in case is useful for you (maybe want to add some other options to the function with some args, as it might be useful in other scenarios).

**Comment by rusty1s** [comment]
Sure, can you send a PR so we can discuss the details?
